### PR TITLE
arch/xtensa: Add special register allocation generator

### PR DESCRIPTION
**Pull request by andyross**
Zephyr likes to use the various Xtensa scratch registers for its own purposes in several places.  Unfortunately, owing to the configurability of the architecture, we have to use different registers for different platforms.  This has been done so far with a collection of different tricks, some... less elegant than others.

Put it all in one place.  This is a python script that emites a "zsr.h" header with register assignments for all the existing users.

[This particular PR is mostly a wash in terms of complexity, removing some hacks and adding an equivalent amount of python and cmake.  But I have plans for this: a similar trick could be played to integrate the long-dormant second-level interrupt handler generator (or better: a rewritten replacement) into the build, so we don't have to have a copy of the output for every platform.  Similar tricks could be played in the link to remove all the boilerplate around the vector table sections (those offsets are all in core-isa.h).  Lots of future here, I promise.]

### CODE REVIEW

**Comment by andyross** [comment]
(Tried to pick a representative set of reviewers from across the zephyr/xtensa community.   Please add anyone I missed.)

**Review by ceolin — APPROVED** [review]

File: arch/xtensa/core/gen_zsr.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+# same on every device.  This script parses a pre-cooked ("gcc -E
+# -dM") core-isa.h file for the current architecture and assigns
+# registers to usages.
+
+NEEDED = ("ALLOCA", "CPU", "FLUSH")
+
+coreisa = sys.argv[1]
+outfile = sys.argv[2]
+
+syms = {}
```
Comment (marc-hb) [inline arch/xtensa/core/gen_zsr.py:19]: Can you do the usual `if __name__ == "__main__":` dance for import+interactive debug?

File: arch/xtensa/core/gen_zsr.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+def get(s):
+    return syms[s] if s in syms else 0
+
+with open(coreisa) as infile:
+    for line in infile.readlines():
+        m = re.match(r"^#define ([^ ]+) ?(.*)", line.rstrip())
```
Comment (marc-hb) [inline arch/xtensa/core/gen_zsr.py:26]: I think this fails to match multiple spaces or tabs:
```
#define       XCHAL...
```
Not a problem thanks to some `-dM` guarantee(s)?

File: arch/xtensa/core/gen_zsr.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        if m:
+            syms[m.group(1)] = 1 if m.group(2) == "" else m.group(2)
```
Comment (marc-hb) [inline arch/xtensa/core/gen_zsr.py:28]: I find this quite cryptic sorry; when is `group(2)` empty and why? Could you add a couple comments or maybe even better: example(s)?

File: arch/xtensa/core/CMakeLists.txt [hunk 4]
```diff
@@ -25,3 +25,26 @@ zephyr_library_sources_ifdef(CONFIG_GDBSTUB gdbstub.c)
...
+# -dM, supported by all Xtensa toolchains) core-isa.h file available
+# as "core-isa-dM.h".  This can be easily parsed by non-C tooling.
+#
+# Note that this adds the SOC/HAL include directory explicitly, they
+# are the official places where we find core-isa.h.
+set(CORE_ISA_DM ${CMAKE_BINARY_DIR}/zephyr/include/generated/core-isa-dM.h)
+set(CORE_ISA_IN ${CMAKE_BINARY_DIR}/zephyr/include/generated/core-isa-dM.c)
+file(WRITE ${CORE_ISA_IN} "#include <xtensa/config/core-isa.h>")
+add_custom_command(OUTPUT ${CORE_ISA_DM}
+  COMMAND ${CMAKE_C_COMPILER} -E -dM
```
[the remark this answers is not in the document]
Comment (andyross) [reply arch/xtensa/core/CMakeLists.txt:39]: It's more than just preprocessing, -dM expands all the recursive definitions and emits a line of output like "#define A xxxx" for every macro it sees.   It's a gcc feature, obviously, but supported just fine on xt-xcc and xt-clang (I checked).

File: arch/xtensa/core/gen_zsr.py (hunk 2 above)
Comment (andyross) [reply arch/xtensa/core/gen_zsr.py:26]: Yeah, the syntax is strict: exactly one space between "#define", the macro name, and its expansion.  It's not like it's specified anywhere in a document though.  Can't hurt to robustify.

File: arch/xtensa/core/gen_zsr.py (hunk 3 above)
Comment (andyross) [reply arch/xtensa/core/gen_zsr.py:28]: The preprocessor behavior of "#define FOO" without a value is to expand FOO to a literal 1.  I don't know why -dM emits it like this, it seems like a wart to me.  Will comment.

File: arch/xtensa/core/gen_zsr.py (hunk 1 above)
Comment (andyross) [reply arch/xtensa/core/gen_zsr.py:19]: Heh, uh... no? :)  Just let this be my one bit of iconoclasm.  I genuinely don't understand why python people think it's a good thing to indent an entire file by an extra tab stop completely needlessly and then put a noop function call at the end of the file.  I mean, functional decomposition in a large script with lots of stuff going on?  Sure.  But this is a trivial straight-through logic kind of thing.  We never did that in shell or perl, I don't see why the culture needs to change here.  As this evolves, we can totally move it in that direction if needed.  For for the logic as it stands, I just like it a lot better like this.

Comment (marc-hb) [reply arch/xtensa/core/gen_zsr.py:19]: > We never did that in shell or perl,

Here are a few technical, non-subjective reasons to do this in shell: https://github.com/thesofproject/sof-test/issues/740

> I genuinely don't understand why python people think it's a good thing to indent an entire file...

Most of the reasons above seem to apply to Python the same. I bet you can find more on the Internet.

I agree it's not important for "small" scripts. The size threshold is of course much more subjective; I feel like this one is very slightly crossing the line. No big deal.

**Review by dcpleung — APPROVED** [review]

File: arch/xtensa/core/gen_zsr.py (hunk 3 above)
Comment (andyross) [reply arch/xtensa/core/gen_zsr.py]: Hm... never mind.  I went back and checked, and in fact all three toolchains do this correctly and emit "#define FOO 1" and not "#define FOO".  Not sure how I got it into my head that they didn't.  Simplified.

Comment (marc-hb) [reply arch/xtensa/core/gen_zsr.py]: BTW Python regex have the useful `[\w]+` and `[\S]+` to match words, clearer and more robust than dealing with whitespace.

https://docs.python.org/3/library/re.html

File: arch/xtensa/core/CMakeLists.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (marc-hb) [reply arch/xtensa/core/CMakeLists.txt:39]: I would expect something like `${CMAKE_CPP}` to expand to something like `gcc -E`. Who wouldn't? :-)

[the remark this answers is not in the document]
Comment (marc-hb) [reply arch/xtensa/core/CMakeLists.txt:39]: `grep` found this in `cmake/dts.cmake`. It seems to be the only place where CPPFLAGS are used.

```
  if(NOT DEFINED CMAKE_DTS_PREPROCESSOR)
    set(CMAKE_DTS_PREPROCESSOR ${CMAKE_C_COMPILER})
  endif()
  ...
  # Run the preprocessor on the DTS input files. We are leaving
  # linemarker directives enabled on purpose. This tells dtlib where
  # each line actually came from, which improves error reporting.
  execute_process(
    COMMAND ${CMAKE_DTS_PREPROCESSOR}
    -x assembler-with-cpp
    -nostdinc
    ${DTS_ROOT_SYSTEM_INCLUDE_DIRS}
    ${DTC_INCLUDE_FLAG_FOR_DTS}  # include the DTS source and overlays
    ${NOSYSDEF_CFLAG}
    -D__DTS__
    ${DTS_EXTRA_CPPFLAGS}
    -E   # Stop after preprocessing
    -MD  # Generate a dependency file as a side-effect
    -MF ${DTS_DEPS}
    -o ${DTS_POST_CPP}
    ${ZEPHYR_BASE}/misc/empty_file.c
```

Most of this code is old though, may predate any fancy toolchain abstraction.

**Review by teburd — APPROVED** [review]

File: arch/xtensa/core/gen_zsr.py (hunk 1 above)
Comment (marc-hb) [reply arch/xtensa/core/gen_zsr.py:19]: > I genuinely don't understand why python people think it's a good thing to indent an entire file...

You just made me wonder why it is customary to have a first level of indentation in all functions in all programming languages. In most languages there is little or no code outside functions so the first indentation level adds no information after all.

Maybe it just subconsciously follows standard typography usage https://en.wikipedia.org/wiki/Margin_(typography)#The_Digital_Page

Off topic sorry.

**Comment by hongshui3000** [comment]
@andyross 
I see some code in the PR that is not dealt with, such as the issue I mentioned.
https://github.com/zephyrproject-rtos/zephyr/issues/40974

**Comment by andyross** [comment]
@hongshui3000 indeed, this is is just cleanup (though it does get the EPC/EPS usage off of the debug interrupt level, which I know was one of your concerns).  I know you have an app architecture with non-Zephyr interrupt handling mixed with Zephyr interrupts.  It's not that we refuse to support that, you just have to realize that it's "unsupported" architecturally and that none of us are likely to fix it for you.  Please consider addressing the issues in an upstreamable way and submitting the fixes yourself.

**Comment by hongshui3000** [comment]
> 

Ok, I see

**Comment by hongshui3000** [comment]
> @hongshui3000 indeed, this is is just cleanup (though it does get the EPC/EPS usage off of the debug interrupt level, which I know was one of your concerns). I know you have an app architecture with non-Zephyr interrupt handling mixed with Zephyr interrupts. It's not that we refuse to support that, you just have to realize that it's "unsupported" architecturally and that none of us are likely to fix it for you. Please consider addressing the issues in an upstreamable way and submitting the fixes yourself.

Ok, I see

File: arch/xtensa/core/gen_zsr.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        maxint = max(il, maxint)
+
+# Now emit our output header with the assignments we chose
+with open(outfile, "w") as f:
+    f.write("/* Generated File, see gen_zsr.py */\n")
+    f.write("#ifndef ZEPHYR_ZSR_H\n")
+    f.write("#define ZEPHYR_ZSR_H\n")
+    for i, need in enumerate(NEEDED):
+        f.write(f"# define ZSR_{need} {regs[i]}\n")
+        f.write(f"# define ZSR_{need}_STR \"{regs[i]}\"\n")
```
Comment (MaureenHelm) [inline arch/xtensa/core/gen_zsr.py:47]: I was curious if any of the NXP platforms needed to be modified in addition to the Intel platforms, so I tried building this PR for `nxp_adsp_imx8`. The generated output looks like this:
```c
/* Generated File, see gen_zsr.py */                                                                
#ifndef ZEPHYR_ZSR_H
#define ZEPHYR_ZSR_H
# define ZSR_ALLOCA MISC0
# define ZSR_ALLOCA_STR "MISC0"
# define ZSR_CPU MISC1
# define ZSR_CPU_STR "MISC1"
# define ZSR_FLUSH EXCSAVE1
# define ZSR_FLUSH_STR "EXCSAVE1"
# define ZSR_EXTRA0 EXCSAVE2
# define ZSR_EXTRA1 EXCSAVE3
# define ZSR_EXTRA2 EXCSAVE4
# define ZSR_RFI_LEVEL 3
# define ZSR_EPC EPC3
# define ZSR_EPS EPS3
#endif
```

Is the space between `#` and `define` deliberate?

Comment (andyross) [reply arch/xtensa/core/gen_zsr.py:47]: Mildly? I wanted to distinguish the macro definitions from the include guards for anyone reading the generated file, but I wasn't willing to dedicate lines in the script to emitting blanks, I guess?

Comment (andyross) [reply arch/xtensa/core/gen_zsr.py:47]: And in theory this is only touching arch-layer code and everyone should pick it up seamlessly, but yeah: it's always possible there's downstream code somewhere already stepping on these specific register assignments that will need to be aware (or potentially integrated).

**Comment by andyross** [comment]
Ping.  Would be good to get this merged.

**Pull request merged**